Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/CosmosServerlessSparkBatchJob.java

```diff
@@ -166,85 +163,77 @@ public String getJobUuid() {
+                                sparkBatchJob.schedulerState().toString(),
+                                getJobLog(sparkBatchJob));
                     }
                 });
     }
 
     @Override
     public Observable<String> awaitStarted() {
-        return getResponsePayloadWithState()
-                .retryWhen(error -> error.flatMap(exception -> {
-                    if (exception instanceof SparkJobFinishedException) {
-                        throw propagate(exception);
-                    } else {
-                        // Retry with limited times
-                        return Observable.just(1)
-                                .zipWith(Observable.range(1, getRetriesMax()), (n, i) -> i)
-                                .delay(getDelaySeconds(), TimeUnit.SECONDS);
+        ctrlInfo("The Spark job is starting...");
+        return getSparkBatchJobRequest()
+                .doOnNext(sparkBatchJob -> {
+                    if (sparkBatchJob.properties() != null && sparkBatchJob.properties().sparkMasterUI() != null) {
+                        ctrlHyperLink(sparkBatchJob.properties().sparkMasterUI() + "?adlaAccountName=" + getAccount().getName());
                     }
-                }))
-                .repeatWhen(ob ->
-                        ob.doOnNext(ignored ->
-                                getCtrlSubject().onNext(
-                                        new AbstractMap.SimpleImmutableEntry<>(Info, "The Spark job is starting...")))
-                                .delay(getDelaySeconds(), TimeUnit.SECONDS))
-                .takeUntil(responsePayload -> isDone(responsePayload.getState()) || isRunning(responsePayload.getState()))
-                .filter(responsePayload -> isDone(responsePayload.getState()) || isRunning(responsePayload.getState()))
-                .flatMap(responsePayload -> {
-                    if (isDone(responsePayload.getState()) && !isSuccess(responsePayload.getState())) {
-                        String errorMsg = "The Spark job failed to start due to "
-                                + String.join("\n", responsePayload.getLog());
-                        log().warn(errorMsg);
-                        return Observable.error(
-                                new SparkJobException(errorMsg));
-                    } else {
-                        return Observable.just(responsePayload.getState());
-                    }
-                });
+                })
+                .flatMap(job ->
+                        getSparkBatchJobRequest()
+                                .repeatWhen(ob -> ob.delay(getDelaySeconds(), TimeUnit.SECONDS))
+                                .takeUntil(sparkBatchJob -> {
+                                    String jobState = getJobState(sparkBatchJob);
+                                    return isJobEnded(sparkBatchJob)
+                                            || (jobState != null && (isDone(jobState) || isRunning(jobState)));
```

Comment:
If `isJobEnded` and `isDone` merged, the code can looks like: `isJobEnded() || isJobStarted()`

Reply:
Yes, these two methods are confusing, I will merge them then.

---
In reply to: [268020311](https://github.com/Microsoft/azure-tools-for-java/pull/2899#discussion_r268020311) [](ancestors = 268020311)